### PR TITLE
Additional S2 support + new S2 boards

### PR DESCRIPTION
**Pull request by UnexpectedMaker**
Extra S2 support (And S3 where applicable):
- PSRAM
- 16MB flash partition
- Added correct ADC and DAC pins
- Added correct GPIO and IRQ pins
- Added S3 area in CMakeLists

Added UM_TinyS2 and UM_FeatherS2 boards and renamed TinyPICO to UM_TinyPICO

### CODE REVIEW

**Comment by mcauser** [comment]
Is it worth moving `dotstar.py` to `/drivers` or `/extmod` so other boards can share it (and include via manifest.py)?

**Comment by UnexpectedMaker** [comment]
The last discussion I had with @dpgeorge about this (a loooooong time ago just as TinyPICO was launching) was that it should probably be move to wherever the other modules live, but that it needed to be made to look more "micropython like" - and I never really got there.

**Comment by UnexpectedMaker** [comment]
> Is it worth moving `dotstar.py` to `/drivers` or `/extmod` so other boards can share it (and include via manifest.py)?

The last discussion I had with @dpgeorge about this (a loooooong time ago just as TinyPICO was launching) was that it should probably be move to wherever the other modules live, but that it needed to be made to look more "micropython like" - and I never really got there.

**Comment by kevinkk525** [comment]
What would be required to make dotstar more micropython like?  Removing the setters?

**Comment by UnexpectedMaker** [comment]
> What would be required to make dotstar more micropython like? Removing the setters and properties?

I honestly cant remember what the conversation entailed, but yeah, I think the properties were some of it. It was a long time ago. It's a small script, so I am sure it will take @dpgeorge < 2min to make it how he wants it to be.

File: ports/esp32/main.c [hunk 1]
```diff
@@ -104,6 +106,20 @@ void mp_task(void *pvParameter) {
             mp_task_heap = malloc(mp_task_heap_size);
             break;
     }
+    #elif CONFIG_ESP32S2_SPIRAM_SUPPORT || CONFIG_ESP32S3_SPIRAM_SUPPORT
+    // Try to use the entire external SPIRAM directly for the heap
+    size_t mp_task_heap_size;
+    size_t esp_spiram_size = esp_spiram_get_size();
+    void *mp_task_heap = (void *)0x3ff80000 - esp_spiram_size;
+    // printf("inside CONFIG_ESP32S2_SPIRAM_SUPPORT ram=%d\n", esp_spiram_size);
```
Comment (dpgeorge) [inline ports/esp32/main.c:114]: best not to leave commented-out lines like this

File: ports/esp32/modmachine.c [hunk 2]
```diff
@@ -39,9 +39,15 @@
 #if CONFIG_IDF_TARGET_ESP32
 #include "esp32/rom/rtc.h"
 #include "esp32/clk.h"
+// #include "driver/touch_pad.h"
```
Comment (dpgeorge) [inline ports/esp32/modmachine.c:42]: why were these includes added with comments?  best to just not have them

File: ports/esp32/boards/sdkconfig.spiram_sx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# MicroPython on ESP32-S2 and ESP32-PAD1_subscript_3, ESP IDF configuration with SPIRAM support
+CONFIG_ESP32S2_SPIRAM_SUPPORT=y
+CONFIG_SPIRAM_TYPE_AUTO=y
+CONFIG_DEFAULT_PSRAM_CLK_IO=30
+CONFIG_DEFAULT_PSRAM_CS_IO=26
+# CONFIG_SPIRAM_FETCH_INSTRUCTIONS is not set
+# CONFIG_SPIRAM_RODATA is not set
+CONFIG_SPIRAM_SPEED_80M=y
+# CONFIG_SPIRAM_SPEED_40M=y
```
Comment (dpgeorge) [inline ports/esp32/boards/sdkconfig.spiram_sx:9]: can these commented-out options be removed?  or did you intend for them to stay?

File: ports/esp32/machine_dac.c [hunk 4]
```diff
@@ -43,12 +43,12 @@ typedef struct _mdac_obj_t {
...
-    #if CONFIG_IDF_TARGET_ESP32
-    {{&machine_dac_type}, GPIO_NUM_25, DAC_CHANNEL_1},
-    {{&machine_dac_type}, GPIO_NUM_26, DAC_CHANNEL_2},
-    #else
+    #if CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3
     {{&machine_dac_type}, GPIO_NUM_17, DAC_CHANNEL_1},
     {{&machine_dac_type}, GPIO_NUM_18, DAC_CHANNEL_2},
+    #else
+    {{&machine_dac_type}, GPIO_NUM_25, DAC_CHANNEL_1},
+    {{&machine_dac_type}, GPIO_NUM_26, DAC_CHANNEL_2},
```
Comment (dpgeorge) [inline ports/esp32/machine_dac.c:51]: this looks like this should have worked before, without the change, no?

File: ports/esp32/machine_pin.c [hunk 5]
```diff
@@ -103,6 +98,90 @@ STATIC const machine_pin_obj_t machine_pin_obj[] = {
...
+    {{&machine_pin_type}, GPIO_NUM_39}, // MTCLK
+    {{&machine_pin_type}, GPIO_NUM_40}, // MTDO
+    {{&machine_pin_type}, GPIO_NUM_41}, // MTDI
+    {{&machine_pin_type}, GPIO_NUM_42}, // MTMS
+    {{&machine_pin_type}, GPIO_NUM_43}, // U0TXD
+    {{&machine_pin_type}, GPIO_NUM_44}, // U0RXD
+    {{&machine_pin_type}, GPIO_NUM_45},
+    {{&machine_pin_type}, GPIO_NUM_46},
+
+#else
```
Comment (dpgeorge) [inline ports/esp32/machine_pin.c:157]: what SoC variant is this bit for?

**Comment by dpgeorge** [comment]
> > What would be required to make dotstar more micropython like? Removing the setters and properties?
> 
> I honestly cant remember what the conversation entailed, but yeah, I think the properties were some of it. It was a long time ago. It's a small script, so I am sure it will take @dpgeorge < 2min to make it how he wants it to be.

I guess you don't want to break backwards compatibility anymore, so that makes it harder to change...

Anyway, since this `dotstar.py` file is moving/renaming I suggest it get moved to `drivers/led/dotstar.py` as-is.  And then in the future someone can make it more portable/generic so it runs on other boards.

**Comment by UnexpectedMaker** [comment]
@dpgeorge I fixed all of those things... some of them (like the machine_pins.c was because I had some C3 stuff in there than I removed, but not cleanly, sorry.
Cheers :)

**Comment by dpgeorge** [comment]
It would be good to have the additions of `UM_TINYS2` and `UM_FEATHERS2` as separate commits.  I can do this during merge, but it would be easier if you could do it @UnexpectedMaker  :)

**Comment by UnexpectedMaker** [comment]
> It would be good to have the additions of `UM_TINYS2` and `UM_FEATHERS2` as separate commits. I can do this during merge, but it would be easier if you could do it @UnexpectedMaker :)

I have no idea how Id do that. Remove the folders from my local repo, re-commit? Then re-add them after the PR is merged?

**Comment by dpgeorge** [comment]
> I have no idea how Id do that.

Something like this (in ports/esp32/ dir):
```
git reset HEAD^^^
git add boards/sdkconfig.spiram_sx partitions-16MiB.csv
git commit -a
(make commit for s2/s3 support)
git add boards/UM_TINYS2/
git commit
(make commit for um-tinys2)
git add boards/UM_FEATHERS2/
git commit
(make commit for um-feathers2)
git push --force
```

**Comment by UnexpectedMaker** [comment]
> > I have no idea how Id do that.
> 
> Something like this (in ports/esp32/ dir):
> 
> ```
> git reset HEAD^^^
> git add boards/sdkconfig.spiram_sx partitions-16MiB.csv
> git commit -a
> (make commit for s2/s3 support)
> git add boards/UM_TINYS2/
> git commit
> (make commit for um-tinys2)
> git add boards/UM_FEATHERS2/
> git commit
> (make commit for um-feathers2)
> git push --force
> ```

Won't that reset ALL of my changes in the PR if I reset to HEAD? My git fu is not strong.

**Comment by dpgeorge** [comment]
> Won't that reset ALL of my changes in the PR if I reset to HEAD? My git fu is not strong.

Yes it'll reset the 3 most recent commits.

You can do it on a new branch:
```
git checkout -b "s2s3-support"
```
then run the commands above

**Comment by dpgeorge** [comment]
See #7222 for a version of this with reworked commits.

**Pull request closed without merging**